### ISSUE TITLE
Local provisioners crash on `command = null`

### ISSUE THREAD

**Issue by KyleKotowick** [hashicorp/terraform#32878]
### Terraform Version

```shell
Terraform v1.4.2
on windows_amd64
```

### Terraform Configuration Files

```terraform
resource "terraform_data" "replacement" {
  input = null
  provisioner "local-exec" {
    when       = create
    on_failure = fail
    command    = null
  }
}
```

### Debug Output

https://gist.github.com/KyleKotowick/a64cfc13578d9eb134951d586d9bc317

### Expected Behavior

Nothing; if `null` is provided for the value of `command`, the provisioner should take no action.

### Actual Behavior

```
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`

### Additional Context

You may ask, "why would you ever have a local provisioner that has a `null` command value?"

Whether the local provisioner is needed is conditional on other factors. We need to be able to do something like;
```terraform
provisioner "local-exec" {
  when       = create
  on_failure = fail
  command    = local.some_condition ? "run.sh" : null
}
```

**Comment by crw** [comment]
Thanks for this report!

**Comment by apparentlymart** [comment]
Hi @KyleKotowick!

Thanks for reporting this indeed!

You are right that the provisioner shouldn't crash in this situation, but `command` is described as required in the documentation so I expect that fixing this bug will mean to replace the panic with a normal error message that says something to the effect of "`command` is required".

If you want the provisioner to "take no action" then I think the closest you could get is to give the provisioner an empty shell script to run: literally `command = ""`. Have you tried that instead of `null` to get the effect you wanted here?

Thanks again!

**Comment by KyleKotowick** [comment]
@apparentlymart Yes, I started with the empty string approach, but it gives this:
```terraform
 Error: Invalid local-exec provisioner command
│
│   with null_resource.cluster[0],
│   on main.tf line 3, in resource "null_resource" "test":
│    3:   provisioner "local-exec" {
│
│ The command must be a non-empty string.
```

You can get around that by using shell commands that do nothing. Our modules use `dirname("/") == "\\" ? "% ':'" : ":"`, which selects a no-op command appropriately for both Windows and Linux systems.

The issue that we're dealing with (and wanting to use `null` for) is performance/speed. Adding a local provisioner is time-costly, which is a problem for modules where the local provisioner is optional/conditional.

For example, this example completes with a `terraform apply auto-approve` in 9 seconds on my laptop:
```terraform
resource "null_resource" "cluster" {
  count = 500
}
```

As soon as I add a local provisioner though, it takes much longer. This example takes 42 seconds:
```terraform
resource "null_resource" "cluster" {
  count = 500
  provisioner "local-exec" {
    when       = create
    on_failure = fail
    interpreter = [
      "powershell.exe"
    ]
    command = dirname("/") == "\\" ? "% ':'" : ":"
  }
}
```

So it increases the per-resource apply time by a factor of 5. That's why I was hoping for a `null` option that would allow me to conditionally run the provisioner.

**Comment by apparentlymart** [comment]
Thanks for trying that and reporting back, @KyleKotowick. I guess the provisioner must have its own validation for the empty string case. Now that I see that, I wouldn't be surprised if it was that validation rule causing the crash you saw, since it's presumably trying to convert the null value into a string before comparing it with the empty string.

Since you are using `null_resource` as a container for your provisioner and `null_resource` doesn't do anything _execept_ act as a container for provisioners, with today's Terraform I'd suggest handling the conditional-enabling of this by making the resource have `count = 0` in situations where you don't wish the provisioner to run at all, since if there aren't any instances of a resource then there is nothing for Terraform to provision and so it won't run the provisioners either.

```hcl
resource "null_resource" "cluster" {
  count = 0

  provisioner "local-exec" {
    when       = create
    on_failure = fail
    interpreter = [
      "powershell.exe"
    ]
    command = dirname("/") == "\\" ? "% ':'" : ":"
  }
}
```

---

If considered as a bug report then the resolution here would be to make it not crash anymore and to return a proper error, since that would be making the `local-exec` provisioner act in the way it was designed to behave.

Adding some way to still declare the provisioner but have it do nothing at all would be an enhancement request because it'd need to be designed first. I understand that you've proposed setting it `command = null` as a way to express that, but because setting an argument to null is the same as omitting it then that would seem to imply making the `command` argument totally optional, so the following would also be treated as valid but do absolutely nothing:

```
  provisioner "local-exec" {
    # nothing at all
  }
```

Perhaps that's okay, but it does seem unfortunate that it would not give any feedback to the user about _why_ it's doing nothing, so perhaps it would work out better to make `command = ""` be the special case that does nothing, so that the provisioner can still treat that argument as required (returning an error if it isn't set) but a module author can explicitly define it as a no-op.

Since [provisioners are a last resort](https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax#provisioners-are-a-last-resort) we tend not to prioritize implementing new features for them, instead focusing on maintaining what's already there so that we can invest more in Terraform's primary features, but perhaps we can make an exception here if `command = ""` seems like a reasonable way to represent "do nothing" since I expect _that_ would be relatively easy to implement at the same time as fixing the crash bug. However, I don't currently have any data on how common it is to need to declare a provisioner that runs but does nothing, so I'm not sure if the additional complexity is justified when it is already possible to factor the provisioner out into a separate `null_resource` and declare zero instances of it, thereby getting the same effect in a way that is generic across all kinds of provisioners, and not just `local-exec`.

**Comment by KyleKotowick** [comment]
@apparentlymart The reason I'm avoiding using the `count` parameter for conditionals is that it immediately adds a requirement for plan-time knowledge of the condition's evaluation. That would break all sorts of applications in our case, since whether you want to run a local provisioner can depend entirely on the output of a resource creation.

`command = ""` would work fine for our case, so I'm a fan of that solution. In terms of "how common" this need is, for context, I came across this issue while working on our fairly popular [Invicton-Labs/shell-resource/external](https://registry.terraform.io/modules/Invicton-Labs/shell-resource/external/latest) modules (67,000 downloads), so I'd say it's a fairly common requirement. Making this change would allow me to update the module to optimize performance on conditional executions, which would affect all users of this module.

If you _do_ make this change, I'd ask that you also make the same change to the [external data source](https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/external) (`program = null` or `program = []` causes it to do nothing). That would allow me to make these same performance improvements to the [Invicton-Labs/shell-data/external](https://registry.terraform.io/modules/Invicton-Labs/shell-data/external/latest) module (another 85,000 downloads).
